presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -272,6 +282,51 @@ public static Slice stGeomFromBinary(@SqlType(VARBINARY) Slice input)
         return serialize(geomFromBinary(input));
     }
 
+    @Description("Converts a Geometry object to a SphericalGeography object")
+    @ScalarFunction("to_spherical_geography")
+    @SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME)
+    public static Slice toSphericalGeography(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        // "every point in input is in range" <=> "the envelope of input is in range"
+        Envelope envelope = deserializeEnvelope(input);
+        if (envelope != null) {
+            checkLatitude(envelope.getYMin());
+            checkLatitude(envelope.getYMax());
+            checkLongitude(envelope.getXMin());
+            checkLongitude(envelope.getXMax());
+        }
+        OGCGeometry geometry = deserialize(input);
+        validateGeometryForSphericalGeography(geometry);
+
+        return input;
+    }
+
+    private static void validateGeometryForSphericalGeography(OGCGeometry geometry)
+    {
+        GeometryType geometryType = GeometryType.getForEsriGeometryType(geometry.geometryType());
+        if (geometryType == GEOMETRY_COLLECTION) {
+            GeometryCollectionIterator iterator = new GeometryCollectionIterator(geometry);
+            while (iterator.hasNext()) {
+                validateGeometryForSphericalGeography(iterator.next());
```

Comment:
@gaoxm Use `geometry.getEsriGeometryCursor()` to avoid recursion. Then, this function will be used only once and the logic could be folded into the caller.

OGCGeometry#is3D is support for geometry collections and therefore, that check can happen once.

Error messages would be clearer if they pointed out that failing operation is conversion to spherical geography.

It is also strange to have check for `geometry.is3D()`, but use `geometry.dimension()` in the type. I also think it is incorrect because a line in a 3D space is still 1-dimensional object.

```
    private static final EnumSet<Type> GEOMETRY_TYPES_FOR_SPHERICAL_GEOGRAPHY =
            EnumSet.of(
                Type.Point,
                Type.Polyline,
                Type.Polygon,
                Type.MultiPoint);
...

        OGCGeometry geometry = deserialize(input);
        if (geometry.is3D()) {
            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Cannot convert 3D geometry to a spherical geography");
        }

        GeometryCursor cursor = geometry.getEsriGeometryCursor();
        while (true) {
            com.esri.core.geometry.Geometry subGeometry = cursor.next();
            if (subGeometry == null) {
                break;
            }

            if (!GEOMETRY_TYPES_FOR_SPHERICAL_GEOGRAPHY.contains(subGeometry.getType())) {
                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Cannot convert geometry of this type to spherical geography: " + subGeometry.getType());
            }
        }

        return input;
```